### PR TITLE
Added fixes for SAM crashes with tests

### PR DESCRIPTION
**Pull request by prenx4x**
*Issue #, if available:*

*Description of changes:*
Fixed some common SAM crashes

*Description of how you validated changes:*
Changes validated using unit tests

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/eventsources/push.py [hunk 1]
```diff
@@ -657,6 +659,15 @@ def _add_swagger_integration(self, api, function, intrinsics_resolver):
                             ),
                         )
 
+                    if not isinstance(method_authorizer, str):
+                        raise InvalidEventException(
+                            self.relative_id,
+                            "Unable to set Authorizer [{authorizer}] on API method [{method}] for path [{path}] "
```
Comment (hoffa) [inline samtranslator/model/eventsources/push.py:665]: More of a general note, but it seems this whole function could benefit from refactoring. It's large, repetitive and consists of multiple intertwined responsibilities (validation and processing).

File: samtranslator/model/eventsources/push.py [hunk 2]
```diff
@@ -218,6 +218,8 @@ class S3(PushEventSource):
     def resources_to_link(self, resources):
         if isinstance(self.Bucket, dict) and "Ref" in self.Bucket:
             bucket_id = self.Bucket["Ref"]
+            if not isinstance(bucket_id, str):
+                raise InvalidEventException(self.relative_id, "'Ref' value in S3 events is not a valid string.")
```
Comment (aahung) [inline samtranslator/model/eventsources/push.py:222]: The error messages seems not match the condition. `bucket_id` here is not a str, while the error message makes me feel like it is a str, just not a valid one. Does the situation here mean the `Ref` key is missing?

Comment (prenx4x) [reply samtranslator/model/eventsources/push.py:222]: Ref is a parameter name which should have some value. like Ref: somevalue . Bucket_id is that value. Therefore we are checking bucket_id which should be a string i.e. value of 'Ref' in the template should be a string. Make sense?

Comment (aahung) [reply samtranslator/model/eventsources/push.py:222]: will `bucket_id` be other types other than `str` and `None`?

File: samtranslator/model/eventsources/push.py (hunk 1 above)
Comment (prenx4x) [reply samtranslator/model/eventsources/push.py:665]: Thats a good point and I totally agree. however it feels like a different task as this refactoring may be needed at several places, not just this function.

File: samtranslator/model/eventsources/push.py (hunk 2 above)
Comment (prenx4x) [reply samtranslator/model/eventsources/push.py:222]: No. I should only be string.

File: samtranslator/model/eventsources/push.py (hunk 1 above)
Comment (hoffa) [reply samtranslator/model/eventsources/push.py:665]: Yep, agreed! Meant it as a general call-out; PR itself looks good (forgot to approve!).

**Review by hoffa — APPROVED** [review]

File: samtranslator/model/eventsources/push.py (hunk 2 above)
Comment (hoffa) [reply samtranslator/model/eventsources/push.py:222]: What about Python 2 where a string can be `unicode`?

Comment (aahung) [reply samtranslator/model/eventsources/push.py:222]: > No. It should only be string.

Comment (aahung) [reply samtranslator/model/eventsources/push.py:222]: > No. It should only be string.

Can I assume if `not isinstance(bucket_id, str)` is `True`, it means the user forgets to put a `Ref`, and as long as `Ref` exists, the `not isinstance(bucket_id, str)` is `False`?

Comment (prenx4x) [reply samtranslator/model/eventsources/push.py:222]: no. The user can have other types like list or dict as well as value for Ref. e.g.
Ref:
  -abc
or
Ref:
  abc:123

Comment (prenx4x) [reply samtranslator/model/eventsources/push.py:222]: we mainly care for hashable values here.

File: samtranslator/model/eventsources/push.py [hunk 3]
```diff
@@ -218,6 +218,8 @@ class S3(PushEventSource):
     def resources_to_link(self, resources):
         if isinstance(self.Bucket, dict) and "Ref" in self.Bucket:
             bucket_id = self.Bucket["Ref"]
+            if not isinstance(bucket_id, str):
```
Comment (hoffa) [inline samtranslator/model/eventsources/push.py:221]: As discussed offline, we should use `six` for string types to deal with Python 2's `unicode` et al.

File: samtranslator/model/eventsources/push.py (hunk 2 above)
Comment (aahung) [reply samtranslator/model/eventsources/push.py:222]: Oh got it, thanks for the explanation!

Comment (aahung) [reply samtranslator/model/eventsources/push.py:222]: resolving this. python 2 related unicode discussion is here https://github.com/aws/serverless-application-model/pull/1757#discussion_r513794030

**Review by aahung — APPROVED** [review]
Except the python 2 unicode concern by @hoffa, LGTM on my side

**Review by hoffa — APPROVED** [review]
Bueno 👌

**Review by jfuss — APPROVED** [review]
LGTM

**Pull request merged**